feather/plugin-host/src/command_dispatcher.rs

```diff
@@ -0,0 +1,163 @@
+use std::{
+    cell::RefCell,
+    convert::TryInto,
+    fmt::{self, Debug},
+    rc::Rc,
+};
+
+use crate::{context::PluginPtrMut, Game, PluginId, PluginManager};
+use anyhow::bail;
+use feather_ecs::{Entity, EntityRef};
+use lieutenant::regex::early_termination::CmdPos;
+use lieutenant::regex::{dfa::DFA, NFA};
+use quill_common::EntityId;
+
+// Constant that should reflect how many commands a default
+// server starts with. Vanilla has roughly 60
+const NUM_COMMANDS: usize = 20;
+
+type CommandId = u32;
+
+/// A collection of data containing what plugin a command was defined in,
+/// the pointer to its Box<Command<...>>, and the regex that determines
+/// when the command should be called.(The command does some parsing beyond the regex).
+pub struct CommandDescription {
+    pub plugin_id: PluginId,
+    pub regex: String,
+    // Should only come from host_call register_system.
+    // it is a pointer to a Box<dyn Command<GameState=(Plugin,CommandContext),z CommandResult=i64>>
+    pub callback: PluginPtrMut<u8>,
+}
+
+impl CommandDescription {
+    // Returns true or false depending on if the command executed or not.
+    // If it returns false that means the commands parser did not recognize
+    // the commnd.
+    /// caller: None => caller was terminal,
+    /// caller: Some(x) => player x called command.
+    fn call(&self, game: &mut Game, input: &str, caller: Option<EntityId>) -> anyhow::Result<()> {
```

Comment:
I personally dislike the idea of having an `Option` to represent whether the terminal or a player was the command sender, as it isn't directly intuitive, and isn't necessarily future-proof if we also consider command blocks. On top of that it might be nice to have a general trait for replying or querying permissions, which could simplify some applications.

Reply:
I agree, @Miro-Andrin therterminal is represented in the ECS and therefore also has an `EntityId`.

Reply:
100% Agree, that is just me being lazy.